## Day 3 Metagenomics Exercise

Todos Santos Genomics and Bioinformatics 2022
---

## First: Install conda environment with software we'll need

Installing bioinformatics software can be challenging!  [Conda](http://anaconda.org) is a tool that facilitate installing software and making sure the installed software works together.  

You will create a new conda environment that includes the tools you'll need for the remainder of this exercise on the thoth01 server.  The tools are listed in [this file](https://github.com/stenglein-lab/2022_Todos_Santos/blob/main/exercises/bio_tools.yaml)

Login to the thoth01 server:
```
ssh your_usernam@thoth01.cvmbs.colostate.edu
```

The username and password are the same ones you've been using for the rna.biology.colostate.edu server.

Create a new conda environment named bio_tools by running this command
```
# download the yaml file from github
curl -OL https://raw.githubusercontent.com/stenglein-lab/2022_Todos_Santos/main/exercises/bio_tools.yaml

# create the conda environment
conda env create --file bio_tools.yaml
```

To activate this conda environment, you will need to run:
```
conda activate bio_tools
```

**You will need to activate this environment every time you want to use these tools.**


## Second: Downloading the data we need

## In this exercise, you will download, process, and evaluate NGS datasets and genome sequences.  You will:

* Download an NGS dataset from the SRA
* Use the FASTQC tool to assess the quality of the reads in the dataset
* Use cutadapt to remove low quality parts of the reads1
* Find and download genome sequences and associated annotation from NCBI

---

### Downloading an SRA dataset

We will work with one of the NGS datasets reported in [this paper](http://journals.plos.org/plospathogens/article?id=10.1371/journal.ppat.1004900)

This dataset was generated by performing shotgun sequencing of total RNA from the liver of a boa constrictor that was diagnosed with [inclusion body disease](https://en.wikipedia.org/wiki/Inclusion_body_disease). The dataset is composed of reads from host RNAs and from viral RNAs.  

<img src="IBD.png" alt="A snake with inclusion body disease twisted up" width="250"/>

To get the dataset, open a browser and navigate to the pubmed page for the dataset's paper:

https://www.ncbi.nlm.nih.gov/pubmed/25993603

Scroll down and find the 'Related information' section of the bottom right of the page.  Click on the SRA link.  This shows the SRA datasets associated with this paper.  Locate the dataset named `shotgun sequencing of tissue from snake_7`.  To download the NGS data, we need to note the accession number (SRR number) for this dataset, which is: SRR1984309.

<img src="snake_7_SRR_number.png" alt="SRR number needed to download SRA datasets" width="400"/>

We're going to download this dataset using the command line tool fastq-dump, part of the [SRA toolkit](https://trace.ncbi.nlm.nih.gov/Traces/sra/sra.cgi?view=toolkit_doc).  First, let's create a directory (folder) in which to work on the thoth01 server.. 

Move to your home directory:
```
# running cd with no options moves you to your home directory
cd
```

Make a new directory
```
mkdir snake_exercise
```

Move to that directory
```
cd snake_exercise
```

Confirm that you are in the correct directory:
```
pwd
```

We will download the dataset using the fastq-dump tool, part of the [SRA toolkit](https://trace.ncbi.nlm.nih.gov/Traces/sra/sra.cgi?view=toolkit_doc). 

To run fasta-dump, you just need to specify the run # (the SRR#) of the dataset you want.  Recall that our run # is SRR1984309. The --split-files option of the command will create 2, synchronized files for the paired reads

```
# download the NGS dataset from SRA run # SRR1984309
fastq-dump --split-files SRR1984309 
```

Confirm that you downloaded the files.  You should see files named SRR1984309_1.fastq and SRR1984309_2.fastq that are each 44 Mb.

```
ls -lh
```

Look at the first 20 lines of the fastq files using the head command
```
head -8 SRR1984309_1.fastq SRR1984309_2.fastq
```

**Questions:**
- What is on each of the 4-lines that make up each fastq sequence?  (See: [FASTQ format](https://en.wikipedia.org/wiki/FASTQ_format))  
- How many reads are in each file?  (Hint: the `wc -l name_of_file` command will tell you the number of lines in the file)

---

### Using FastQC to evaluate quality of NGS data

[FastQC](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/) is a tool that: 

> ... aims to provide a simple way to do some quality control checks on raw sequence data coming from high throughput sequencing pipelines. It provides a modular set of analyses which you can use to give a quick impression of whether your data has any problems of which you should be aware before doing any further analysis

Performing a quality check like this of your data is one of the first things you'll want to do when you receive your new sequencing data (or when you download a dataset from an online repository like the SRA).

FastQC can be used via a graphical interface or via the command line. On the command line, run fastq to perform a quality check on these datasets: 

```
fastqc SRR1984309_1.fastq SRR1984309_2.fastq
```

These datasets have already been pre-cleaned, so they look pretty good.  Note that there is possible Nextera adapter contamination towards the end of some reads.  This makes sense, because the libraries were made with the Nextera protocol.  In the next section, we will trim those off.

---

### Read trimming with cutadapt

[fastp](https://github.com/OpenGene/fastp) is a tool that can be used to trim low quality and adapter sequences from NGS reads.  It is essential to trim adapter sequences and low quality bases off of raw NGS reads.

We will run this command to trim our reads:

```
fastp -i SRR1984309_1.fastq -I SRR1984309_2.fastq -o SRR1984309_1_trimmed.fastq -O SRR1984309_2_trimmed.fastq --length_required 100
``` 

fastp automatically detects and removes adapter sequences from NGS reads.

Here, we are discarding reads shorter than 100 bases after trimming.

```
ls -lh
```

Now, we can use fastqc to analyze the trimmed datasets:
```
fastqc SRR1984309_1_trimmed.fastq SRR1984309_2_trimmed.fastq
```

Transfer the fastq html files using sftp to your computer and open them html files in a browswer and answer these questions:

- How many reads remain after trimming?
- Did the quality of the basecalls improve?
- Did the trimming remove Nextera adapters?

Note: There are many trimming tools. Other popular trimming tools include [BBDuk](https://jgi.doe.gov/data-and-tools/software-tools/bbtools/bb-tools-user-guide/bbduk-guide/) and [cutadapt](https://cutadapt.readthedocs.io/en/stable/).

---

### Downloading the boa constrictor genome.

The dataset we downloaded was created by sequencing a library made from boa constrictor liver RNA.  We will map the reads in this dataset to the boa constrictor genome sequence.

First, we need to *find* the boa constrictor genome.  As usual, there are few ways we could go about this:

1. navigate through the NCBI [Taxonomy database](https://www.ncbi.nlm.nih.gov/taxonomy/)
2. navigate through the NCBI [Genome database](https://www.ncbi.nlm.nih.gov/genome/)
3. navigate through another genome database, like [Ensembl](http://www.ensembl.org/index.html) or [UCSC](https://genome.ucsc.edu/) 
4. google 'boa constrictor genome sequence'  (not a terrible way to do it)

- Let's choose option 1, and go through the NCBI Taxonomy database.  Navigate to https://www.ncbi.nlm.nih.gov/taxonomy/
   - Search for `boa constrictor`.  
   - Click on Boa constrictor link, then click the Boa constrictor link again
   - You should see a table in the upper right corner showing linked records in various NCBI (Entrez) databases.
   - Click on the `Genome (1)` link in that table to go to the boa constrictor records in the NCBI Genome database 

   - The linked page should say 'No items found', because the boa constrictor genome isn't actually in the NCBI Genome database<sup>[1](#myfootnote1)</sup>.  However, there is a **mitochondrial** genome. Let's download that.
   - Click on "See also 1 organelle- and plasmid-only records matching your search"
   - In the Replicon Info table, note the link to the boa constrictor mtDNA genome sequence (NC_007398.1)
   - Click on this 'NC_007398.1' RefSeq link

Now we need to download the sequence.  We'll do this through the browser.  In the upper right hand corner of the page, note the 'Send' drop down 

- Click Send->Complete Record->File->Format[FASTA]->Create File

This will download this file in fasta format to your laptop.  You should have downloaded a fasta file of ~19 kb, named sequence.fasta, or something like that.  

Next, download the sequence in GenBank format too.  Note that this file is larger (~42 kb), because it contains annotation as well as the actual sequence.

Note that the downloaded files have unhelpful names: `sequence.fasta` and `sequence.gb` or similar.  Rename these files to NC_007398.fasta and NC_007398.gb

Use a text editor (Notepad++ or BBEdit or something similar) to inspect the files you've just downloaded.  Note that the fasta file does not have any annotation, but the genbank file does.  As we will see, both of these formats will be useful.

Use Cyberduck or FileZilla to transfer these files to the server and move them to the snake_exercise directory.

At this point, you should be on the thoth01 server in the snake_exercise folder and there should be a file named `NC_007398.fasta`
```
# should be in snake_exercise folder
pwd

# should see a 19 kb file named NC_007398.fasta
ls -lh 
```

## Map reads to the boa constrictor mitochondrial genome

### Create a bowtie index from the boa constrictor mitochondrial genome sequence

Read mapping tools map reads very quickly because they use pre-built indexes of the reference sequence(s).  We'll use the [Bowtie2](http://www.nature.com/nmeth/journal/v9/n4/full/nmeth.1923.html) mapper. Bowtie2 has a nice [manual](http://bowtie-bio.sourceforge.net/bowtie2/manual.shtml) that explains how to use this software. 

There are a variety of other good read mapping tools, such as [BWA](https://github.com/lh3/bwa) and [minimap2](https://github.com/lh3/minimap2), which works well for long read data or long sequences.

The first step will be to create an index of our reference sequence (the boa constrictor mitochondrial genome).

Now, we'll use the bowtie2-build indexing program to create the index.  This command takes 2 arguments: 
(1) the name of the fasta file containing the sequence(s) you will index
(2) the name of the index (can be whatever you want)

```
# create a bowtie2 index.  Name it NC_007398_bt_index
bowtie2-build NC_007398.fasta NC_007398_bt_index 
```

Confirm that you have successfuly built the index.  Bowtie2 indexes are actually composed of 6 files.  You should see six files with names ending in bt2, like NC_007398_bt_index.3.bt2
```
# should see six bowtie2 index files
ls -lh
```

Note that this index building went very fast for a small genome like the boa mtDNA, but can take much longer (hours) for Gb-sized genomes.  But at least you only need to build an index once per reference sequence.

### Mapping reads in the SRA dataset to the boa constrictor mitochondrial genome 

Now that we've created the index, we can map reads to the boa mtDNA.  We'll map our fastp-trimmed paired reads to this sequence, as follows:

```
bowtie2 -x NC_007398_bt_index \
   -1 SRR1984309_1_trimmed.fastq \
   -2 SRR1984309_2_trimmed.fastq \
   --no-unal \
   --threads 8 \
   -S SRR1984309_mapped_to_NC_007398.sam
```

Let's deconstruct this command line:

Option | Meaning
--- | ---
-x NC_007398_bt_index   |  -x: name of index you created with bowtie2-build
-1 SRR1984309_1_trimmed.fastq   | name of the paired-read FASTQ file 1
-2 SRR1984309_2_trimmed.fastq   | name of the paired-read FASTQ file 2
--no-unal   | don't output unmapped reads to the SAM output file (this will make the sam file _much_ smaller)
--threads 8   | since this server has multiple processers, run on 8 processors to go faster
-S SRR1984309_mapped_to_NC_007398.sam   | name of output file in SAM format

Bowtie will output some information about what percentage of the reads aligned.
 - What percentage of the reads aligned to the boa constrictor mitochondrial genome?
 - Does that number make sense?

The main output file from bowtie (and other read mappers) is a file in [sam format](https://en.wikipedia.org/wiki/SAM_(file_format)), which describes how reads aligned to the reference sequence.  The output file from our command is `SRR1984309_mapped_to_NC_007398.sam`

Sam is a plain text format, so you can look at the first 20 lines by running this command:

```
head -20 SRR1984309_mapped_to_NC_007398.sam
```

You can see that there are several header lines beginning with `@`, and then one line for each mapped read.  See [here](http://genome.sph.umich.edu/wiki/SAM) or [here](https://samtools.github.io/hts-specs/SAMv1.pdf) for more information about interpreting SAM files.

Answer the following questions about the first mapped read:
- What position in the mtDNA sequence did it map to?
- What is the mapping quality for this read's mapping?
- What does this mapping quality score indicate?  


### Visualizing aligned (mapped) reads in Geneious

Geneious provides a nice graphical interface for visualizing the aligned reads described in your SAM file.   Other tools for visualizing this kind of data include [IGV](http://software.broadinstitute.org/software/igv/) and [Tablet](https://ics.hutton.ac.uk/tablet/)

First, you need to transfer the SRR1984309_mapped_to_NC_007398.sam file from thoth01 to your computer.  Use sftp or cyberduck to do this.

Second, you need to have your reference sequence in Geneious, preferably with annotations.  You can do this 2 ways:

1. Drag and drop the NC_007398.gb file into a folder in Geneious
2. Download the file directly into Genious, using the NCBI->Nucleotide interface (search for NC_007398.1).  Once downloaded, drag from the NCBI download folder into another folder in Geneious.  

Once you have the boa constrictor mitochondrial genome in a folder in Geneious, you can drag and drop the SAM file that bowtie2 output into the same folder.  Geneious will tell you that it 'can't find the sequence it needs in the selected file'.  It is telling you it is trying to find the reference sequence to which you aligned reads.  Answer: 'Find a sequence with the same name in this Geneious folder' or 'Use one of the selected sequences' (after selecting the boa mtDNA sequence).

- A few Geneious tips:
  - Enlarge the Geneious window so that it fills the screen
  - Click View->Expand Document View to enlarge the alignment
  - Try playing with the visualization settings in the panels on the right of the alignment

- Some questions to consider when viewing the alignment:
  - What is the average coverage depth across the mitochondrial genome?
  - Is the coverage even across the mitochondrial genome?
  - Would you expect coverage to be even across the genome?  (Recall that this data is derived from total RNA from liver tissue). 
  - Are the mitochondrial genes expressed evenly?
  - Are there any variants between this snake's mitochondrial genome sequence and the boa constrictor reference sequence?
  - Is it expected that there are variants between these reads and this reference sequence?  Explain your answer.
  - Can you distinguish true variants from sequencing errors?
  - In general, how can you distinguish true variants from sequencing errors?
  - Is it possible that reads that derive from the boa constrictor nuclear genome are mapping to this sequence?
  - How would you prevent nuclear reads from mapping to the mitochondrial genome?
  - Can you identify mapped read pairs?


---
Footnotes

<a name="myfootnote1">1</a>: The boa constrictor genome _was_ sequenced as part of the [Assemblathon 2 competition](https://www.ncbi.nlm.nih.gov/pubmed/23870653), and the (unannotated) assemblies can be found [here](http://gigadb.org/dataset/100060), which you can find by Google searching "boa constrictor genome".  So Google searching is not a bad idea.

